### PR TITLE
Redshift/line list optimization and minor fixes

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request (essentially a continuation of #1031) refactors the line list lines to update instead of being destroyed and re-drawn, as well as a number of other small tweaks aimed at improving the experience using the redshift slider:
* tweak slider resolution to prioritize "smooth" sliding with a handful of shown lines (with 1000 steps over the range of the x-limits), while minimizing "catch-up time" for a large number of lines (via dynamic throttling of the slider based on the number of drawn lines, with a linear scale to roughly try to achieve a throttle that is the same as the time it takes to redraw that number of lines)
* apply comparable resolution to redshift/RV input elements as slider so those also react to the x-limits and the manual `range` option in `set_redshift_slider_bounds`
* dynamic rounding of reported redshift precision when using slider or +/- buttons (to have no more than the number of digits in the current step size - removes issues with floating point precision with repeated slider use)
* fix clearing lines from plot when deleting line list
* fix lines when changing y-limits or x/y-units of spectrum viewer
* moves API helper methods from Specviz helper to LineListMixin
* add LineListMixin to Mosviz/Specviz2d (to access `mosviz.load_spectral_lines`, `mosviz.set_redshift_slider_bounds` etc).
* update plotted lines when loading line list from API
* fix bug where loading line list from API clears previously loaded lists in the UI 
* remove entries from `spectral_lines` when deleting list from UI
* support for spectral equivalencies during unit conversion.
* plotting new line no longer resets previously plotted lines to rest wavelength [:cat:](https://jira.stsci.edu/browse/JDAT-2092) (may require revisiting if supporting per-line/list redshifts)

**POSTPONED TO FUTURE WORK**
- determine support for per-line or per-line-list redshift support [:cat:](https://jira.stsci.edu/browse/JDAT-2079)
- determine how to handle redshifts from custom loaded table (see AAS workshop notebook and #475 - right now these are used on load, but overridden as soon as the line list plugin is opened and are then replaced by the "global" redshift)
- export redshifts in `mosviz.spectral_line_list` (currently keep the originally loaded values, but aren't replaced by "global" redshift) [:cat:](https://jira.stsci.edu/browse/JDAT-2080)
- Add support for creating a new line that requires equivalencies [:cat:](https://jira.stsci.edu/browse/JDAT-2103)

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: docs/specviz/redshift.rst
Comment (pllim) [inline docs/specviz/redshift.rst:17]: It is less clear now that you are supposed to access this method in Specviz, hence the suggestion.
```suggestion
:meth:`~jdaviz.configs.default.plugins.line_lists.line_list_mixin.LineListMixin.set_redshift_slider_bounds`
method in ``specviz`` by specifying the ``range`` and/or ``step`` keywords.  By setting either or both
```

File: jdaviz/configs/default/plugins/line_lists/line_list_mixin.py
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_list_mixin.py:65]:
```suggestion
        for default values based on the limits of the spectrum plot.
```

File: jdaviz/configs/default/plugins/line_lists/line_list_mixin.py
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_list_mixin.py:71]:
```suggestion
            Note that the slider specifies delta redshift from the current value, so a
```

File: jdaviz/configs/default/plugins/line_lists/line_list_mixin.py
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_list_mixin.py:94]: Redshift is float, right? Do you want to use `np.allclose` here instead or do you really require exact floating point match?

File: jdaviz/configs/default/plugins/line_lists/line_lists.py
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_lists.py:35]: Should we start using `IntHandleEmpty` and `FloatHandleEmpty` from #1041 or is that not applicable here?

File: jdaviz/configs/default/plugins/line_lists/line_lists.py
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_lists.py:202]: Here, you are invoking Quantity constructor on every update, even if `rs_redshift` has not changed. If performance is your goal, you might want to change it to Quantity somewhere else.

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (pllim) [inline jdaviz/configs/specviz/plugins/viewers.py:231]: Why this change?

File: jdaviz/core/marks.py
Comment (pllim) [inline jdaviz/core/marks.py:5]: Total nitpick on my part. 😸 
```suggestion
from astropy import units as u
from bqplot.marks import Lines
from glue.core import HubListener
from specutils import Spectrum1D
```

File: jdaviz/core/marks.py
Comment (pllim) [inline jdaviz/core/marks.py:46]:
```suggestion
        x = old_quant.to_value(x_all.unit, equivalencies=u.spectral())
```

File: jdaviz/core/marks.py
Comment (pllim) [inline jdaviz/core/marks.py:90]: What about wavenumber?

File: jdaviz/core/marks.py
Comment (pllim) [inline jdaviz/core/marks.py:101]:
```suggestion
        self._rest_value = old_quant.to_value(new_unit, equivalencies=u.spectral())
```

**Review by pllim — COMMENTED** [review]
Just a general review for now. I feel like someone more familiar with Specviz and maybe a PO should review this. But if you need me to step in as second review, please ping me.

Do you plan to add tests?

File: jdaviz/configs/default/plugins/line_lists/line_lists.py
Comment (kecnry) [reply jdaviz/configs/default/plugins/line_lists/line_lists.py:35]: that shouldn't be applicable here since these traitlets are not exposed as inputs to the user and are only set internally or through the API.

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (kecnry) [reply jdaviz/configs/specviz/plugins/viewers.py:231]: I suspect this was caused by a missed change while rebasing... I'll look into that and revert once I confirm if that was the case.

Comment (kecnry) [reply jdaviz/configs/specviz/plugins/viewers.py:231]: although `scales` was intentionally removed as that logic is now handled by the `viewer/self` reference in the `SpectralLine` object itself.

**Comment by kecnry** [comment]
@pllim - thanks for the review/suggesetions!  The existing test coverage reaches the requirements, but I'll try to see if I can come up with any additional tests to prevent regression wrt any of the fixed bugs.

File: docs/mosviz/plugins.rst
Comment (kecnry) [reply docs/mosviz/plugins.rst:51]: Your suggestion is definitely more concise... but do you think it's still clear enough that the lag is referring to the movement of the spectral lines, and not the slider itself?  If that is still clear, I'll adopt your suggestion.  Otherwise maybe something like: "If the shifting of spectral lines lag behind the slider, try plotting less lines"?

Comment (pllim) [reply docs/mosviz/plugins.rst:51]: Yeah, I like "If the shifting of spectral lines lag behind the slider, try plotting less lines." Thanks!

**Comment by pllim** [comment]
> The existing test coverage reaches the requirements

So, the redshift stuff is already tested? I am just a little surprised that you didn't need to update any tests even though you refactored a bunch of stuff.

File: jdaviz/core/marks.py
Comment (kecnry) [reply jdaviz/core/marks.py]: see c663cc7

File: jdaviz/core/helpers.py
Comment (kecnry) [inline jdaviz/core/helpers.py:49]: TODO: rebase to make use of #1054 if/once merged

**Review by javerbukh — APPROVED** [review]
Looks good! It's really impressive how smoothly it's running considering how many marks are on the screen!

File: CHANGES.rst
Comment (pllim) [inline CHANGES.rst:73]:
```suggestion
- Spectral lines and redshift are refactored to improve performance. [#1036]
```

**Review by pllim — APPROVED** [review]
LGTM.

Maybe future work: I noticed that manual entry of redshift breaks if I start typing the float with period instead of leading zero, i.e., `.0001`.

**Comment by kecnry** [comment]
@pllim 

> Maybe future work: I noticed that manual entry of redshift breaks if I start typing the float with period instead of leading zero, i.e., .0001.

I now handle never raising errors for the empty `'.'` which handles cases like `.123`, but `.01` will still cause issues as the `.0` is immediately parsed as 0 and overwrites the decimal.  This likely occurs everywhere we have a float input across the app, so I think deferring is reasonable. 

If you're happy with this change for now, I'll go ahead and merge and create a follow-up ticket for the float input with leading decimal.

**Pull request merged**
